internal/controller/flux/plugindefinition_controller.go

```diff
@@ -0,0 +1,123 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Greenhouse contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package flux
+
+import (
+	"context"
+	"time"
+
+	"golang.org/x/time/rate"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	sourcecontroller "github.com/fluxcd/source-controller/api/v1"
+
+	greenhousev1alpha1 "github.com/cloudoperators/greenhouse/api/v1alpha1"
+	"github.com/cloudoperators/greenhouse/internal/clientutil"
+	"github.com/cloudoperators/greenhouse/internal/controller/cluster/utils"
+)
+
+const (
+	defautlNameSpace         = "greenhouse"
+	pluginDefinitionUrlField = "spec.url"
+)
+
+// PluginDefinitionReconciler reconciles plugindefinitions and translates them into Flux resources
+type PluginDefinitionReconciler struct {
+	client.Client
+}
+
+// Greenhouse related RBAC rules for the FluxReconciler
+//+kubebuilder:rbac:groups=greenhouse.sap,resources=plugindefinitions,verbs=get;list;watch
+
+// Flux related RBAC rules for the FluxReconciler
+// +kubebuilder:rbac:groups=source.toolkit.fluxcd.io,resources=helmrepositories,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=source.toolkit.fluxcd.io,resources=helmrepositories/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=source.toolkit.fluxcd.io,resources=helmrepositories/finalizers,verbs=get;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=events,verbs=create;patch
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *PluginDefinitionReconciler) SetupWithManager(name string, mgr ctrl.Manager) error {
+	r.Client = mgr.GetClient()
+
+	// index PluginDefinitions by the URL field for faster lookups
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &sourcecontroller.HelmRepository{}, pluginDefinitionUrlField, func(rawObj client.Object) []string {
+		helmRepository, ok := rawObj.(*sourcecontroller.HelmRepository)
+		if helmRepository.Spec.URL == "" || !ok {
+			return nil
+		}
+		return []string{helmRepository.Spec.URL}
+	}); err != nil {
+		return err
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(controller.Options{
+			RateLimiter: workqueue.NewTypedMaxOfRateLimiter(
+				workqueue.NewTypedItemExponentialFailureRateLimiter[reconcile.Request](30*time.Second, 1*time.Hour),
+				&workqueue.TypedBucketRateLimiter[reconcile.Request]{Limiter: rate.NewLimiter(rate.Limit(10), 100)}),
+			MaxConcurrentReconciles: 3,
+		}).
```

Comment:
Same as for the plugin_controller no need to add additional rate limiting. Let's first see if we require this